### PR TITLE
perf: optimize spark_base64 in spark-expr

### PR DESCRIPTION
**Pull request by andygrove**
## What changed
Optimizes `spark_base64` in the native `spark-expr` crate.

Reuse scratch String buffers for base64 encode + line-wrap and build directly into a preallocated StringBuilder, eliminating two per-row heap allocations in the array path.

## Evidence
- Correctness: unit tests + seeded differential fuzz (bit-identical Arrow output vs `main`).
- Benchmark (criterion, Rust-only): see numbers below; gate required at least one benchmark ≥ 5.0% faster (non-overlapping confidence intervals) and no benchmark meaningfully slower.

```
spark_base64_ long, chunked: 21.073% faster (base 882215ns -> cand 696308ns)
spark_base64_ short, chunked: 11.932% faster (base 216853ns -> cand 190979ns)
spark_base64_ short, unchunked: 12.069% faster (base 215962ns -> cand 189896ns)
spark_base64_ long, unchunked: 11.43% faster (base 627480ns -> cand 555761ns)
```

### CODE REVIEW

**Review by mbutrovich — COMMENTED** [review]
The scratch-buffer reuse is the real win and it stands on its own. One concern with the capacity precompute in `encode_array`:

```rust
let data_capacity: usize = (0..array.len())
    .filter(|&i| array.is_valid(i))
    .map(|i| base64_encoded_len(array.value(i).len()))
    .sum();
```

Two issues:

1. This adds a full O(n) pass over the array before the encode loop makes its own pass. For a change whose goal is removing per-row allocations, the extra walk is a cost that partly offsets the win.

2. `base64_encoded_len` returns the unwrapped length. When `chunk` is true (the Spark default), the output includes CRLF separators (about 2 bytes per 76 chars), so the builder under-reserves and can still reallocate on long chunked input. The PR description says it "sizes the value buffer up front", which only holds for the unchunked path.

I would drop the precompute entirely. The builder's amortized doubling is fine, and removing it saves the extra pass. If you want to keep it, make the estimate chunk-aware so the reservation is actually tight for the default path.

Separately, the Evidence block in this PR is clean (all four reported benchmarks exist in the added `base64.rs`), which is worth noting since the sibling perf PRs had reported numbers that do not map to committed benchmarks.

File: native/spark-expr/src/string_funcs/base64.rs [hunk 1]
```diff
@@ -83,12 +115,19 @@ fn encode(bytes: &[u8], chunk: bool) -> String {
...
+    chunk_into_lines_buf(&encoded, &mut out);
+    out
+}
+
+/// Append the CRLF-wrapped form of `encoded` (lines of at most 76 characters) into `out`.
+/// The caller is expected to have cleared `out`; only used when `encoded.len() > LINE_LEN`.
+fn chunk_into_lines_buf(encoded: &str, out: &mut String) {
     let separators = (encoded.len() - 1) / LINE_LEN;
-    let mut out = String::with_capacity(encoded.len() + separators * 2);
+    out.reserve(encoded.len() + separators * 2);
```
Comment (comphead) [inline native/spark-expr/src/string_funcs/base64.rs]: should we reserve every row? perhaps we can do a reusable buffer?

File: native/spark-expr/src/string_funcs/base64.rs
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs:167]: `native/spark-expr/src/string_funcs/base64.rs:74-102` is the code this PR rewrites, yet both existing tests (`base64.rs:147` and `base64.rs:157`) only call `encode()`, which is the scalar path. `encode()` is unchanged by this PR. Nothing exercises the new builder loop, the null branch (`base64.rs:87-89`), the capacity pass, the `LargeBinary` (`i64`) generic instantiation, or the array-level wrap condition. The differential fuzz gate covers this at runtime, but a regression here would not be caught by `cargo test`.

Suggested change: add a test that builds a `BinaryArray` and a `LargeBinaryArray` mixing nulls, empty values, a value that encodes to exactly 76 chars, and a value that wraps twice, then asserts the resulting `StringArray` element-for-element (including a null in the middle). Reuse the `[b'a'; 57]` / `[b'c'; 120]` fixtures already in the chunked test so the expected wrapping is known.

File: native/spark-expr/src/string_funcs/base64.rs [hunk 2]
```diff
@@ -62,11 +64,41 @@ pub fn spark_base64(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionE
...
-        .map(|value| value.map(|bytes| encode(bytes, chunk)))
-        .collect()
+    // Encode into a builder, reusing scratch buffers across rows. This avoids a
+    // fresh per-row heap allocation for each element's base64 string (and, when
+    // chunking, its line-wrapped copy) that the previous per-element
+    // `encode()` + `collect()` incurred, and sizes the value buffer up front.
+    let data_capacity: usize = (0..array.len())
+        .filter(|&i| array.is_valid(i))
+        .map(|i| base64_encoded_len(array.value(i).len()))
+        .sum();
```
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs]: `base64.rs:79-82` iterates the whole array once to sum encoded lengths, then the main loop (`base64.rs:86-100`) iterates again. Each `array.value(i)` call in the capacity pass reconstructs a slice. For the chunked long case the sum also omits the CRLF bytes, so the value buffer still reallocates during wrapping, which is exactly the "long, chunked" benchmark row.

Suggested change: get the total input byte count in O(1) from `array.value_data().len()` and derive a single upper bound, for example `base64_encoded_len(total_bytes)` plus, when `chunk`, a CRLF slack term. This drops the extra per-row pass and stops under-sizing the wrapped output. If you keep the per-row exact sum, at minimum add the separator bytes into `data_capacity` when `chunk` is true.

File: native/spark-expr/src/string_funcs/base64.rs [hunk 3]
```diff
@@ -62,11 +64,41 @@ pub fn spark_base64(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionE
     }
 }
 
+const LINE_LEN: usize = 76;
```
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs:68]: `LINE_LEN` itself is single-sourced at module scope (`base64.rs:67`), so the threshold value cannot drift. What lives in two places is the decision to wrap: the array path inlines `chunk && encoded.len() > LINE_LEN` (`base64.rs:93`) while the scalar path relies on `chunk_into_lines`'s internal `encoded.len() <= LINE_LEN` guard (`base64.rs:118`). The two are consistent today (both leave a 76-char string unwrapped), but the comparison lives in two places and a future edit to one can silently diverge from the other.

Suggested change: have the array path call a single shared helper that owns the threshold, for example `maybe_chunk(&encoded, chunk, &mut wrapped)` that returns which buffer to append, so `chunk_into_lines` and the array loop cannot drift. This also removes the near-duplicate wrapper `chunk_into_lines` / `chunk_into_lines_buf` split.

File: native/spark-expr/src/string_funcs/base64.rs [hunk 4]
```diff
@@ -83,12 +115,19 @@ fn encode(bytes: &[u8], chunk: bool) -> String {
...
-    const LINE_LEN: usize = 76;
     if encoded.len() <= LINE_LEN {
         return encoded;
     }
+    let mut out = String::new();
+    chunk_into_lines_buf(&encoded, &mut out);
+    out
+}
+
+/// Append the CRLF-wrapped form of `encoded` (lines of at most 76 characters) into `out`.
```
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs]: `base64.rs` doc comment on `chunk_into_lines_buf` says "The caller is expected to have cleared `out`; only used when `encoded.len() > LINE_LEN`." The function itself is correct for any input (it reserves and appends), so the "only used when" clause is a usage note, not a precondition, and the "expected to have cleared" wording reads like an unchecked invariant. State why the buffer is reused (to avoid per-row allocation) rather than describing current call sites.

Suggested change: shorten to "Appends the CRLF-wrapped form of `encoded` to `out`. `out` is reused across rows to avoid a per-row allocation, so the caller clears it before each call."

**Review by mbutrovich — CHANGES_REQUESTED** [review]
Another round, thanks @andygrove!

File: native/spark-expr/src/string_funcs/base64.rs (hunk 1 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs]: The scratch buffer is now reused across rows (see \`encode_array\` — one \`buf\` outside the loop), so we only pay for a resize on rows that outgrow the previous row.

File: native/spark-expr/src/string_funcs/base64.rs
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs:167]: Added `encode_array_binary_chunked` (BinaryArray with mixed nulls, empty value, exactly-76 boundary, wrap-once, wrap-twice) and `encode_array_large_binary_unchunked` covering the LargeBinaryArray i64 offset instantiation.

File: native/spark-expr/src/string_funcs/base64.rs (hunk 2 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs]: Replaced the O(n) pre-pass with an O(1) estimate: `base64_encoded_len(array.value_data().len())`, plus `chunked_len(...)` slack when `chunk` is true. The wrapped-output long-input path no longer under-reserves.

File: native/spark-expr/src/string_funcs/base64.rs (hunk 3 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs:68]: Consolidated into a single `encode_into(bytes, chunk, out)` helper. The threshold decision lives in one place; the scalar `encode` and array `encode_array` both call it. Removed the near-duplicate `chunk_into_lines`/`chunk_into_lines_buf` split.

File: native/spark-expr/src/string_funcs/base64.rs (hunk 4 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs]: Rewrote the doc on the encoding helper: `out is reused across rows to avoid per-row heap allocations; the caller clears it before each call` (no more `only used when` clause that reads like a precondition).

File: native/spark-expr/src/string_funcs/base64.rs [hunk 5]
```diff
@@ -62,33 +64,43 @@ pub fn spark_base64(args: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionE
...
+    let unwrapped_len = base64_encoded_len(bytes.len());
+    if unwrapped_len <= LINE_LEN {
+        BASE64_STANDARD.encode_string(bytes, out);
+        return;
+    }
+    // Reuse `out` for the wrapped result: encode into a temporary owned by the outer scratch,
+    // then copy CRLF-wrapped chunks in. The temporary is short-lived per row, but the caller's
+    // long-lived scratch avoids the per-row allocation the previous implementation had.
+    let mut encoded = String::with_capacity(unwrapped_len);
+    BASE64_STANDARD.encode_string(bytes, &mut encoded);
```
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs]: This allocates a fresh `String` on every call whose encoding exceeds 76 chars, which is the entire long/chunked benchmark row. The comment on line 100 ("the caller's long-lived scratch avoids the per-row allocation the previous implementation had") overstates the result: the caller's `buf` is reused, but this temporary reintroduces the allocation for exactly the case the buffer reuse was meant to remove.

MIME wrapping is aligned to base64's block structure: 57 input bytes encode to exactly 76 base64 chars with no padding (57 is divisible by 3, `19 * 4 = 76`). So you can encode the input in 57-byte windows straight into `out`, with `\r\n` between windows, and drop the `encoded` temporary and the second pass entirely:

```rust
let mut first = true;
for window in bytes.chunks(57) {
    if !first {
        out.push_str("\r\n");
    }
    BASE64_STANDARD.encode_string(window, out);
    first = false;
}
```

One pass, zero per-row allocation, same output as the current `chunk_into_lines` semantics.

File: native/spark-expr/src/string_funcs/base64.rs [hunk 6]
```diff
@@ -98,12 +110,43 @@ fn chunk_into_lines(encoded: String) -> String {
...
     }
+}
+
+fn encode_array<O: OffsetSizeTrait>(array: &GenericBinaryArray<O>, chunk: bool) -> StringArray {
+    // Right-size the value buffer in O(1) from the input's total byte count. When chunking,
+    // add the CRLF slack the encoded output will contain so the long-input path does not grow.
+    let total_bytes = array.value_data().len();
+    let encoded_total = base64_encoded_len(total_bytes);
+    let data_capacity = if chunk {
+        chunked_len(encoded_total)
```
Comment (mbutrovich) [inline native/spark-expr/src/string_funcs/base64.rs]: `base64_encoded_len` is applied to the summed raw bytes, but each row pads independently to a multiple of 4. For N rows of 1 byte the estimate is about `4N/3` against an actual `4N`, so it reserves roughly a third of what is needed, and the chunked case also misses the per-row CRLFs. This is only a capacity hint (the builder grows correctly), so no correctness impact, but the "sizes the value buffer up front" claim only holds for the few-large-row shape. If you take the 57-byte-window approach above, encode straight into the builder's value buffer and this estimate can stay as a rough lower bound without the extra care.

**Review by mbutrovich — CHANGES_REQUESTED** [review]
Another round, thanks @andygrove. All four earlier threads are resolved: the array/null/LargeBinary/wrap paths now have direct coverage (`encode_array_binary_chunked`, `encode_array_large_binary_unchunked`), the O(n) pre-pass is gone, the wrap threshold is single-sourced in `encode_into`, and the doc comment now states the reuse rationale.

One new issue from the rewrite, plus a minor follow-on.

File: native/spark-expr/src/string_funcs/base64.rs (hunk 5 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs]: You are right that the temporary reintroduces the allocation for exactly the case the reuse was meant to remove, and the doc comment overstated the result. Both fixed in b1acaefbc — but not via 57-byte windows, because I benchmarked that and it is slower.

I implemented your suggestion verbatim first. It is correct (57 is divisible by 3, `19 * 4 == 76`, and only the final window can carry padding, so the concatenation is byte-identical to encode-then-split — I added an exhaustive test for that below). But it regresses the case it targets:

| Benchmark | 57-byte windows |
| --- | --- |
| short, unchunked | -2.5% |
| short, chunked | **+3.7%** |
| long, unchunked | -6.1% |
| long, chunked | **+10.5%** |

The per-call overhead of many small `encode_string` calls (each recomputes the encoded length and reserves before delegating to `encode_slice`) costs more than the single bulk encode plus the copy. A 200-byte value is 4 windows, and base64 encoding is fast enough in bulk that call overhead dominates at that size. The allocation you flagged is real but is not the dominant cost — a ~270-byte `String` is a cheap thread-cached malloc.

So I kept the bulk encode and removed the allocation a different way: `encode_into` now writes to a `fmt::Write` sink and takes a caller-owned `scratch` for the unwrapped encoding. `GenericStringBuilder` implements `fmt::Write`, so the array path passes **the builder itself** as the sink and the CRLF-wrapped result lands directly in its value buffer. That removes the per-row temporary *and* the copy that previously staged each row in `buf` before `append_value`:

| Benchmark | `fmt::Write` into builder |
| --- | --- |
| short, unchunked | ~0% |
| short, chunked | -1.7% |
| long, unchunked | -7.2% |
| long, chunked | **-20.8%** |

Same baseline, same machine. So the batch now performs no per-row allocation, which was your point, and long chunked input got 21% faster rather than 10% slower.

I kept your window insight documented on `encode_into`, including why it is not used, so nobody re-derives it and re-regresses the benchmark. The exhaustive equivalence test (`chunked_matches_encode_then_split_at_every_length`, sweeping every length across three line boundaries) also stays, since it pins the line structure regardless of which strategy is used.

File: native/spark-expr/src/string_funcs/base64.rs (hunk 6 above)
Comment (andygrove) [reply native/spark-expr/src/string_funcs/base64.rs]: Correct on all counts — the estimate was not a bound at all for arrays of short values, and the "sizes the value buffer up front" claim only held for the few-large-row shape. Fixed in b1acaefbc.

Since rows pad independently, the encoded total is `sum ceil(len_i / 3) * 4`, which genuinely cannot be recovered from the summed input bytes. But `sum ceil(x_i) <= ceil(sum x_i) + (N - 1)`, so adding `4 * (N - 1)` turns it into a true upper bound while staying O(1) — no pass over the offsets, so this does not walk back the earlier round. It over-reserves by at most 4 bytes per row, which only bites for arrays of very short values, where the buffer is small in absolute terms anyway. The chunked case then applies `chunked_len` to that bound, which is still an upper bound because `sum floor((enc_i - 1) / 76) <= floor((sum enc_i - 1) / 76)`.

Extracted into `encoded_capacity` with the reasoning in a doc comment, and pinned by `encoded_capacity_is_an_upper_bound`, which compares the estimate against the actual encoded total for eight shapes — including the `vec![1; 64]` worst case that motivated this, all-empty, exactly-one-line, wraps-once, mixed, and few-large-rows — under both chunk settings. `encode_array_many_tiny_rows` covers output correctness for that shape separately.

On encoding straight into the builder's value buffer: that turned out to be possible and is now done (see the other thread), which is also why this estimate is back to being a pure hint rather than something the hot path depends on.

**Comment by andygrove** [comment]
@mbutrovich both threads addressed in b1acaefbc, with details inline. Short version:

**The temporary `String` is gone, and the fix went further than removing it.** I implemented your 57-byte-window suggestion first and benchmarked it — it is correct, but it is **+10.5%** on long chunked input, the case it targets, because the per-call overhead of many small `encode_string` calls exceeds the single bulk encode plus the copy. So instead: `encode_into` now writes to a `fmt::Write` sink with a caller-owned scratch, and since `GenericStringBuilder` implements `fmt::Write`, the array path passes **the builder itself** — the wrapped output lands directly in the value buffer. That removes the per-row temporary *and* the `buf` → builder copy that both the old code and the window version pay:

| Benchmark | 57-byte windows | `fmt::Write` into builder |
| --- | --- | --- |
| short, unchunked | -2.5% | ~0% |
| short, chunked | +3.7% | -1.7% |
| long, unchunked | -6.1% | -7.2% |
| long, chunked | **+10.5%** | **-20.8%** |

Your insight about the 57-byte block alignment is documented on `encode_into` along with why it is not used, so it does not get re-derived and re-regressed later.

**Capacity is now a real upper bound.** `sum ceil(x_i) <= ceil(sum x_i) + (N - 1)`, so `+ 4 * (N - 1)` bounds the per-row padding while staying O(1) — no reintroduced pass over the offsets. Extracted to `encoded_capacity` with the derivation in a doc comment.

**Tests:** an exhaustive sweep over every input length across three line boundaries (asserting equality with encode-then-split, that every non-final line is exactly 76 chars, and that none exceeds it), a capacity upper-bound check over eight array shapes including the all-one-byte worst case under both chunk settings, and an all-tiny-rows array test.

Verified: 517 crate unit tests pass, `cargo clippy --all-targets` clean, and the three base64 SQL file tests (`base64.sql`, `base64_unchunked.sql`, `unbase64.sql`) pass against real Spark on 3.5 — those cover the CRLF-wrapped values and the `unbase64` round trip, so the wrapping is validated against Java's MIME encoder end to end and not just against my own reimplementation of it.

**Review by mbutrovich — APPROVED** [review]
Approved pending CI, thanks @andygrove!
